File: Work/LeetCode/Problems/249-GroupShiftedStrings/GroupShiftedStrings.cpp
/*************************************************************************
    > File Name: GroupShiftedStrings.cpp
    > Author: lwei2
    > Mail: lwei2@IT.com 
    > The Description :
 ************************************************************************/


#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution
{
	public:
		vector<vector<string> >groupStrings(vector<string> &strings)
		{
			vector<vector<string>> ans;
			unordered_my<string, multiset<string>> m;
			for(auto a :  strings)
			{
				string t = "";
				for(char c : a)
				{
					t = t + to_string((c-a[0] + 26)%26) + ",";
				}
				m[t].insert(a);
			}
			for(auto it = m.begin(); it != m.end(); it++)
				ans.push_back(vector<string>(it->second.begin(), it->second.end()));
			return ans;
		}
};

int main(void)
{
	return 0;
}

File: Work/LeetCode/Problems/249-GroupShiftedStrings/GroupShiftedStrings.java
/*************************************************************************
    > File Name: GroupShiftedStrings.java
    > Author: lwei2
    > Mail: lwei2@IT.com 
    > The Description :
 ************************************************************************/

public List<List<String> >groupStrings2(String[] strings)
{
	List<List<String>> res = new ArrayList<List<String>>();
	HashMap<String, ArrayList<String>> m = new HashMap<String, ArrayList<String>>();
	for(String a : strings)
	{
		StringBuilder  sb = new StringBuilder();
		for(char c : a.toCharArray())
		{
			sb.append((c + 26 - a.charAt(0))%26);
			sb.append(",");
		}
		String t = sb.toString();
		if(m.containsKey(t))
			m.get(t).add(a);
		else
		{
			m.put(t, new ArrayList<>());
			m.get(t).add(a);
		}
	}
	res.addAll(m.values());
	return res;
}
